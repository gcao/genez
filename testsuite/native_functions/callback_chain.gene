#!/usr/bin/env gene run

# Test: Gene -> Zig -> Gene -> Zig callback chains

# Define a Gene function that uses native functions
(fn process_with_logging [x]
  (var doubled = (* x 2))
  # This will call native print
  (print "  Processing:" x "-> doubled:" doubled)
  doubled)

# Test callback chain
(print "Testing callback chain with map:")
(var result = (map [1 2 3] process_with_logging))
(print "Result:" result)

# More complex chain: filter with logging
(fn check_and_log [x]
  (var is_big = (> x 5))
  (print "  Checking:" x "-> is_big:" is_big)
  is_big)

(print "\nTesting filter with logging:")
(var filtered = (filter [3 6 9 2 8] check_and_log))
(print "Filtered result:" filtered)

# Reduce with accumulator logging
(fn add_with_log [acc x]
  (var sum = (+ acc x))
  (print "  Accumulator:" acc "+ " x "=" sum)
  sum)

(print "\nTesting reduce with logging:")
(var total = (reduce [1 2 3 4] add_with_log 0))
(print "Total:" total)