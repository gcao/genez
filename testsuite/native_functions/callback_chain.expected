Testing callback chain with map:
  Processing: 1 -> doubled: 2
  Processing: 2 -> doubled: 4
  Processing: 3 -> doubled: 6
Result: [2 4 6]
\nTesting filter with logging:
  Checking: 3 -> is_big: false
  Checking: 6 -> is_big: true
  Checking: 9 -> is_big: true
  Checking: 2 -> is_big: false
  Checking: 8 -> is_big: true
  Checking: 3 -> is_big: false
  Checking: 6 -> is_big: true
  Checking: 9 -> is_big: true
  Checking: 2 -> is_big: false
  Checking: 8 -> is_big: true
Filtered result: [6 9 8]
\nTesting reduce with logging:
  Accumulator: 0 +  1 = 1
  Accumulator: 1 +  2 = 3
  Accumulator: 3 +  3 = 6
  Accumulator: 6 +  4 = 10
Total: 10